vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+    VTR_ASSERT(valid_node_id(node));
+
+    node_capacities_[node] = capacity;
+}
+
+void RRGraph::set_node_ptc_num(RRNodeId node, short ptc) {
+    VTR_ASSERT(valid_node_id(node));
+
+    node_ptc_nums_[node] = ptc;
+}
+
+void RRGraph::set_node_pin_num(RRNodeId node, short pin_id) {
+    VTR_ASSERT(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == IPIN || node_type(node) == OPIN, "Pin number valid only for IPIN/OPIN RR nodes");
+
+    set_node_ptc_num(node, pin_id);
+}
+
+void RRGraph::set_node_track_num(RRNodeId node, short track_id) {
+    VTR_ASSERT(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == CHANX || node_type(node) == CHANY, "Track number valid only for CHANX/CHANY RR nodes");
+
+    set_node_ptc_num(node, track_id);
+}
+
+void RRGraph::set_node_class_num(RRNodeId node, short class_id) {
+    VTR_ASSERT(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == SOURCE || node_type(node) == SINK, "Class number valid only for SOURCE/SINK RR nodes");
+
+    set_node_ptc_num(node, class_id);
+}
+
+void RRGraph::set_node_cost_index(RRNodeId node, short cost_index) {
+    VTR_ASSERT(valid_node_id(node));
+    node_cost_indices_[node] = cost_index;
+}
+
+void RRGraph::set_node_direction(RRNodeId node, e_direction direction) {
+    VTR_ASSERT(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == CHANX || node_type(node) == CHANY, "Direct can only be specified on CHANX/CNAY rr nodes");
+
+    node_directions_[node] = direction;
+}
+
+void RRGraph::set_node_side(RRNodeId node, e_side side) {
+    VTR_ASSERT(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == IPIN || node_type(node) == OPIN, "Side can only be specified on IPIN/OPIN rr nodes");
+
+    node_sides_[node] = side;
+}
+
+void RRGraph::set_node_R(RRNodeId node, float R) {
+    VTR_ASSERT(valid_node_id(node));
+
+    node_Rs_[node] = R;
+}
+
+void RRGraph::set_node_C(RRNodeId node, float C) {
+    VTR_ASSERT(valid_node_id(node));
+
+    node_Cs_[node] = C;
+}
+
+/*
+ * Set a segment id for a node in rr_graph 
+ */
+void RRGraph::set_node_segment(RRNodeId node, RRSegmentId segment_id) {
+    VTR_ASSERT(valid_node_id(node));
+
+    /* Only CHANX and CHANY requires a valid segment id */
+    if ((CHANX == node_type(node))
+        || (CHANY == node_type(node))) {
+        VTR_ASSERT(valid_segment_id(segment_id));
+    }
+
+    node_segments_[node] = segment_id;
+
+    return;
+}
+
+/* For a given node in a rr_graph
+ * classify the edges of each node to be configurable (1st part) and non-configurable (2nd part) 
+ */
+void RRGraph::partition_node_in_edges(RRNodeId node) {
+    //Partition the edges so the first set of edges are all configurable, and the later are not
+    auto first_non_config_edge = std::partition(node_in_edges_[node].begin(), node_in_edges_[node].end(),
+                                                [&](const RREdgeId edge) { return edge_is_configurable(edge); }); /* Condition to partition edges */
```

Comment:
I think this can just be:
```
auto first_non_config_edge = std::partition(node_in_edges_[node].begin(), node_in_edges_[node].end(), edge_is_configurable)?
```
(no need for the lambda).

Reply:
I have tried avoid use lambda. But complier complains due to the edge_is_configurable is a const function.